### PR TITLE
Always order required actions by priority (regardless of context)

### LINKED ISSUE #16873: Required actions execution order (session and user required actions)
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

authentication

### Describe the bug

Hi, when I configured some required actions like so:

- password reset (priority 1)
- verify email (priority 2)
- terms and conditions (priority 3)

When the user triggers a password reset (via forgot password) I expect the following order:

action |context
--------|-------
password reset| session
verify email |user
terms and conditions |user

but instead they execute in the following order:

action        | context
--------|-------
verify email |user
terms and conditions |user
password reset| session

This seems like a bug to me and I looked into the code and it seems to go
wrong  [here](https://github.com/keycloak/keycloak/blob/9b01bf382d69b0eb4f3d0defd27beddbac15ba37/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java#L1160-L1166)

```java
Stream<String> requiredActions = user.getRequiredActionsStream();
// Here all the user required actions are executed, even if there are higher priority actions stored in the session...
Response action = executionActions(session, authSession, request, event, realm, user, requiredActions);
if (action != null) return action;

//... which are executed afterwards
action = executionActions(session, authSession, request, event, realm, user, authSession.getRequiredActions().stream());
if (action != null) return action;

```

One potential way of fixing this would be to concatenate user required actions and session required actions to execute them in one step.

The only obvious difference would be that we can't return early after executing the user actions, but I don't understand when that is actually desired.

### Version

19.0.2

### Expected behavior

When configuring required actions like so:

- password reset (priority 1)
- verify email (priority 2)
- terms and conditions (priority 3)

When the user triggers a password reset (via forgot password) I expect the following order:

action |context
--------|-------
password reset| session
verify email |user
terms and conditions |user

### Actual behavior

but instead they execute in the following order:

action        | context
--------|-------
verify email |user
terms and conditions |user
password reset| session

### How to Reproduce?

This can be reproduced by configuring some required actions like in the example above.

### Anything else?

Previously reported and discussed over here:

- https://groups.google.com/g/keycloak-dev/c/MqpYd7kKa3U
- https://keycloak.discourse.group/t/required-actions-execution-order-session-and-user-required-actions/10689

### PR DESCRIPTION
**Pull request by danielFesenmeyer**
- AuthenticationManager#actionRequired: make sure that the highest prioritized required action is performed first, possibly before the currently requested required action
- AuthenticationManager#nextRequiredAction: make sure that the next action is requested via URL, also based on highest priority (-> requested URL will match actually performed action, unless required actions for the user are changed by a parallel operation)
- add tests to RequiredActionPriorityTest, add helper method for priority setup to ApiUtil (for easier and more robust setup than up-to-now)

Closes #16873

### CODE REVIEW

**Comment by danielFesenmeyer** [comment]
Draft PR for solving the linked issue. Currently draft only, because there are some (already) existing tests to be fixed - or maybe further existing code to be adjusted.

**Comment by mposolda** [comment]
@thomasdarimont Could you please review this PR? Is it aligned with your prototype in the PR https://github.com/keycloak/keycloak/pull/28419 ?

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java [hunk 1]
```diff
@@ -1033,45 +1034,40 @@ public static Response finishedRequiredActions(KeycloakSession session, Authenti
...
-        }
-        if (!authSession.getRequiredActions().isEmpty()) {
-            return authSession.getRequiredActions().iterator().next();
-        }
-
-        String kcAction = authSession.getClientNote(Constants.KC_ACTION);
-        if (kcAction != null) {
-            return kcAction;
+        final var kcAction = authSession.getClientNote(Constants.KC_ACTION);
+        final var firstApplicableAction =
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:1046]: Just a single nitpick: how about the more generic "nextApplicable"?.

**Review by thomasdarimont — COMMENTED** [review]
LGTM Just one minor nitpick about naming that we could change.

I think the rest is pretty solid! Great work @danielFesenmeyer!

I haven't tried it, but I just want to confirm: What happens when a required action is explicitly targeted via the `kc_action` parameter? Will the other actions still be ignored as before? Same question for executing an action via E-Mail link.

**Comment by danielFesenmeyer** [comment]
@thomasdarimont Thanks for your feedback. I'll have a deeper look at the `kc_action` handling. There are still a few tests failing, probably related to this topic.

**Comment by danielFesenmeyer** [comment]
My code still had a bug regarding the "cancel-AIA" behavior: the latest draft now considers that an application initiated action (AIA) cannot be cancelled.

All tests - except one - are succeeding now without further adjustments. The only test to adjust was WebAuthnRegisterAndLoginTest#webAuthnPasswordlessAlternativeWithWebAuthnAndPassword:
The test uses a flow which sets required actions (`WebAuthnPasswordlessRegisterFactory` and `WebAuthnRegisterFactory`) into the session, in case there is no registration of credentials yet. Additionally, it explicitly sets the required action `WebAuthnPasswordlessRegisterFactory` on user level. With the old behavior, user context required actions were executed before session context required actions --> `WebAuthnPasswordlessRegisterFactory` before `WebAuthnRegisterFactory`. Now the execution is based on priority: `WebAuthnRegisterFactory` (prio 70) is now executed before `WebAuthnPasswordlessRegisterFactory` (prio 80).
Do you think this small behavioral change is ok?

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java (hunk 1 above)
Comment (danielFesenmeyer) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java]: I agree, that's a better name - adjusted.

**Comment by thomasdarimont** [comment]
Awesome thanks @danielFesenmeyer for the great work!

I'm unsure about the behavior change and I'd like to get the opinion of @mposolda for this.

**Review by mposolda — DISMISSED** [review]
@danielFesenmeyer @thomasdarimont Thanks for this PR and feedback! To me, the behaviour change is ok. The new behaviour is much better IMO.

**Comment by mposolda** [comment]
@danielFesenmeyer The thing is, that this branch has conflicts. Can you please rebase and fix the conflicts? I hope to merge this PR afterwards.

**Comment by danielFesenmeyer** [comment]
@mposolda Rebased - new flaky test is unrelated to the changes from my POV.

**Review by mposolda — APPROVED** [review]
@danielFesenmeyer Thanks!

**Pull request merged**